### PR TITLE
feat(app): add bash.exec basic tool support

### LINKED ISSUE #637: [Feature]: Shell command governance: Bash execution and Bash rule parsing support
# Shell Command Governance: Bash Execution and Bash Rule Parsing Support

## Summary

This issue defines a shell-governance proposal for LoongClaw.

It is a direction-setting issue, not an implementation plan. The goal is to
make one thing clear:

> How do we add a minimal, governable approval model for real Bash workflows
> without bypassing LoongClaw's existing capability -> policy -> execution ->
> audit chain?

## Background

LoongClaw already has an explicit tool-governance path:

```text
CapabilityToken -> PolicyEngine.authorize(...) -> PolicyExtensionChain -> Execution -> Audit
```

So this proposal is not about creating a shell side-path. It is about extending
the existing path so it can govern real shell workflows more reliably than the
current `shell.exec` model.

Today, `shell.exec` is much closer to "spawn a program with arguments" than to
"understand and govern a shell command." That is workable for simple commands,
but it breaks down for common workflows such as:

- `cd foo && cargo test`
- `cargo fmt ; cargo test`
- `cargo test || cargo test -- --nocapture`
- `source .venv/bin/activate`

The real problem is not "support every shell feature now." The real problem is:

> how to support Bash execution and Bash rule parsing with a minimal governance
> model that remains conservative when the structure is unclear.

## Why Bash First

This issue starts with Bash because it is the most practical convergence point,
not because Bash is the intended long-term boundary.

The tradeoff is straightforward:

- Bash-style workflows cover a large share of day-to-day development tasks.
- Based on the public implementation evidence reviewed, Rust already has a
  realistic Bash AST baseline for a "parse conservatively, then review" path.
- Pulling Bash, PowerShell, and broader multi-shell abstraction into the same
  proposal would blur the scope again.

This proposal therefore takes the following approach:

- make the Bash path work first
- make minimal-command-unit splitting work first
- make a small user-facing rule surface work first
- leave `shell.exec` redirection and broader shell support to later work

### 1. Add `bash.exec` as the experimental entry

This proposal should add `bash.exec` without changing the existing `shell.exec`
behavior yet.

The point of `bash.exec` is to give us an isolated experimental surface:

- it keeps the new Bash path independent while it is still maturing
- it avoids changing current `shell.exec` semantics too early
- it leaves room to redirect `shell.exec` to the same approval chain later

`bash.exec` is therefore the carrier for this proposal, not a permanent product
boundary.

### 2. Keep the request surface small

The initial `bash.exec` surface should stay minimal:

- `command`
- optional `cwd`
- optional timeout

This proposal does not try to introduce:

- cross-call shell state
- persistent shell sessions
- env-governance semantics
- PTY / interactive execution

In other words, execution is single-shot for now.

### 3. Use Bash parsing internally, not as a public DSL

The parser's job in this proposal is narrow:

- conservatively split a shell command into minimal command units
- identify structures that the system cannot safely govern yet

The parser is not being exposed as a user-facing AST rule language.

That distinction matters. A parser is useful immediately for safer command
understanding. A public AST DSL would turn this issue into a much larger
language-design project.

## Minimal Command Unit Model

The system should attempt to split a compound Bash command into minimal command
units and then review each unit independently.

Example:

- `cd foo && cargo test`
  - `cd foo`
  - `cargo test`

Here, a "minimal command unit" means:

> a command-level unit that can independently enter rule evaluation and that
> the system can understand with enough confidence to govern conservatively

This model also applies to commands with conditional execution.

For example:

- `cargo test || cargo test -- --nocapture`

The right rule is not "only review what will definitely execute." The rule is:

> if a command unit may execute, it must be covered by either an explicit rule
> or the configured default policy

So `||` is not excluded because the parser cannot read it. It is handled
conservatively because both sides have execution potential.

### What is in scope now

This proposal only commits to a small user-facing rule surface:

- `prefix_rule(...)`

The initial user-facing rule model is intentionally simple:

- explicit `deny`
- explicit `allow`
- fallback to `Default`

For example:

```python
prefix_rule(
    prefix = ["cargo", "test"],
    decision = "allow",
    note = "verification command",
)

prefix_rule(
    prefix = ["cargo", "publish"],
    decision = "deny",
    note = "release path is blocked for now",
)
```

The current preferred direction is to use a declarative rule-file substrate
based on Starlark, while keeping LoongClaw's own semantics and governance
model.

The goal is not Codex compatibility. The goal is to reuse a practical
declarative format without implying that LoongClaw shares Codex's execution or
sandbox behavior.

### Why only prefix rules

Prefix rules are enough to unlock a meaningful first step:

- they cover a real set of command-governance scenarios
- they keep the user model understandable
- they avoid prematurely committing to regex, wildcard, or structural AST
  predicates

This is a deliberate scope control choice.

## Decision Flow

This proposal preserves the repository's current default semantics:

1. `Deny`
2. `Allow`
3. `Default`

`Default` currently means `Allow` or `Deny`, depending on configuration. This
matches the current `shell.exec` behavior in the repository.

`Ask` remains a longer-term direction, but it is not treated here as already
shipped infrastructure.

For each minimal command unit, the system should evaluate:

1. explicit deny rules
2. explicit allow rules
3. configured default mode

Then the whole command is aggregated as follows:

1. if any unit resolves to `Deny`, the whole command is `Deny`
2. otherwise the whole command resolves according to the remaining unit results
   and `Default`

In practice:

- with `Default = Allow`, unsupported or unknown structures push toward allow
- with `Default = Deny`, unsupported or unknown structures push toward deny

## Conservative Fallback

This proposal does not claim fine-grained governance for every Bash structure.

If a command:

- cannot be split reliably, or
- contains structures that this proposal does not govern precisely yet

the system should not pretend it understood the command well enough. It should
fall back to `Default`.

That conservative fallback is a core part of the model.

## Structures Not Given Fine-Grained Semantics Yet

Even if the parser can recognize them, the following are not being promised as
fine-grained rule surfaces in this proposal:

- environment-variable prefix assignments
- redirection
- pipelines
- subshells
- command substitution

Examples:

- `FOO=1 cargo test`
- `cargo test > out.txt`
- `cargo test | tee out.txt`
- `echo $(git rev-parse HEAD)`

This is not a parser limitation claim. It is a scope claim.

### Linux / macOS

Expose `bash.exec` normally when Bash is available.

### Windows

Attempt to support a compatible Bash runtime on Windows.

If no acceptable Bash runtime is available:

1. do not expose `bash.exec`
2. continue to rely on the existing `shell.exec`
3. emit an explicit warning or log such as `bash unavailable`

This must not fail silently.

## Non-Goals

This proposal does not commit to:

- a full sandbox redesign
- a broader capability/security model rewrite
- a public AST predicate DSL
- fine-grained env semantics
- fine-grained redirection / pipeline / substitution permissions
- session-scoped remember UX
- full parallel multi-shell support
- a fixed long-term convergence shape between `bash.exec` and `shell.exec`

## Acceptance Criteria

This direction is successful if an implementation can satisfy at least the
following:

1. `bash.exec` exists as the experimental Bash-governance entry point
2. `bash.exec` still runs through capability -> policy -> execution -> audit
3. the system can conservatively split a meaningful subset of Bash compound
   commands
4. understood minimal command units can be reviewed through `prefix_rule(...)`
5. users can persist `allow` / `deny` prefix rules in a rule file
6. unsupported or unresolved cases fall back to `Default(Allow|Deny)`
7. complex structures are not misrepresented as already having fine-grained
   governance support
8. on Windows, `bash.exec` is not exposed silently when Bash is unavailable;
   the system keeps the current `shell.exec` path and emits an explicit signal

## Design Constraints

This issue is intentionally not an implementation checklist, but it should
still preserve the constraints that define the proposal boundary:

1. keep `bash.exec` separate from `shell.exec` while the model is still being
   proven
2. define minimal command units conservatively from Bash structure rather than
   from raw strings
3. route decisions through the existing policy chain shape instead of inventing
   a shell-only approval path
4. treat unsupported or only-partially-understood structures as `Default`,
   rather than inventing optimistic partial semantics
5. keep the first user rule surface small enough that richer rule types
   can extend it without rewriting the first model

An implementation following this proposal will still need to answer:

1. which Bash AST forms are accepted as governable minimal command units
2. how those units map into `prefix_rule(...)` evaluation
3. how rule loading and precedence attach to the existing policy extension flow
4. how `Default(Allow|Deny)` is surfaced without implying shipped `Ask`
   infrastructure
5. how Windows runtime detection exposes or withholds `bash.exec` while keeping
   failure visible

## Possible Later Extensions

Later work may build on this proposal in a cleaner order:

- richer rule types
- explicit `Ask` infrastructure
- finer shell-structure governance
- `shell.exec` redirection into the same chain
- broader shell support

Those are follow-up directions, not part of the proposal boundary defined here.

### PR DESCRIPTION
**Pull request by ThreeIce**
## Summary

- Problem:
  - The current PR needs final release-readiness cleanup so the shipped `bash.exec` AST governance slice is documented and validated accurately.
- Why it matters:
  - Reviewers need an honest record of what this PR changes, what stays out of scope, and which remaining red checks belong to unrelated pre-existing baseline issues.
- What changed:
  - Added `bash.exec` as an experimental parallel tool without replacing `shell.exec`.
  - Added AST-based minimal-command-unit governance for `bash.exec`, including conservative fallback behavior and deny precedence.
  - Added Starlark-authored prefix-rule loading plus compatibility inputs from legacy `shell_allow`, `shell_deny`, and `shell_default_mode`.
  - Aligned default Bash rules to `~/.loongclaw/rules`, refreshed the April architecture-drift artifact, added explicit `cargo deny` exceptions for the `starlark` transitive unmaintained advisories, and updated user-facing `bash.exec` docs.
- What did not change (scope boundary):
  - `shell.exec` remains unchanged.
  - This PR does not add `approval_required` / `approve_always` behavior for `bash.exec`.
  - This PR does not fix discovery follow-up / `tool.search` model behavior; that follow-up is tracked separately in `#678`.
  - This PR does not replace the Starlark rules parser.
  - This PR does not fix the pre-existing strict docs baseline issues or the post-merge `tools/mod.rs` architecture-budget breach; both are explicitly left for separate follow-up PRs because they were not introduced by this PR.

## Linked Issues

- Closes #637
- Related #678

## Change Type

- [x] Bug fix
- [x] Feature
- [ ] Refactor
- [x] Documentation
- [x] Security hardening
- [x] CI / workflow / release

## Touched Areas

- [ ] Kernel / policy / approvals
- [ ] Contracts / protocol / spec
- [ ] Daemon / CLI / install
- [ ] Providers / routing
- [x] Tools
- [ ] Browser automation
- [ ] Channels / integrations
- [x] ACP / conversation / session runtime
- [ ] Memory / context assembly
- [x] Config / migration / onboarding
- [x] Docs / contributor workflow
- [x] CI / release / workflows

## Risk Track

- [ ] Track A (routine / low-risk)
- [x] Track B (higher-risk / policy-impacting)

If Track B, fill these in:

- Risk notes:
  - `bash.exec` now enforces AST-governed allow/deny evaluation and consumes local rules plus legacy shell-policy compatibility inputs.
  - Governance defaults matter: unmatched / unsupported / parse-unreliable units can now deny execution under default-deny.
- Rollout / guardrails:
  - `bash.exec` stays experimental and parallel to `shell.exec`.
  - Broken rule files fail closed and are hidden from discovery.
  - Discovery follow-up prompt behavior remains explicitly out of scope and tracked separately in `#678`.
- Rollback path:
  - Revert the `bash.exec` governance series and this cleanup commit to return to the earlier basic-tool slice.

## Validation

- [x] `cargo fmt --all -- --check`
- [x] `cargo clippy --workspace --all-targets --all-features -- -D warnings`
- [x] `cargo test --workspace --locked`
- [x] `cargo test --workspace --all-features --locked`
- [ ] Relevant architecture / dep-graph / docs checks for touched areas
- [x] Additional scenario, benchmark, or manual checks when behavior changed
- [x] If this changes config/env fallback, limits, or defaults: include before/after behavior and regression coverage for explicit path, fallback path, and boundary values
- [x] If tests mutate process-global env: document how state is restored or serialized

Commands and evidence:

```text
cargo fmt --all -- --check
cargo clippy --workspace --all-targets --all-features -- -D warnings
cargo test --workspace --locked
cargo test --workspace --all-features --locked
cargo audit
cargo deny check advisories bans sources
bash scripts/check_architecture_drift_freshness.sh docs/releases/architecture-drift-2026-04.md
git diff --check

Environment note for the final test re-run:
- The Windows-backed C: volume became full during local verification, so the final pre-commit `cargo test` reruns used `CARGO_TARGET_DIR=/tmp/loongclaw-target-ci-parity` to keep build artifacts off the full volume. This was a local-only verification workaround and did not change repository contents or PR behavior.

Manual bash.exec smoke evidence:
- `printf ok`: allowed when `bash.exec` actually executed
- `git status --short`: allowed when `bash.exec` actually executed
- `git rev-parse HEAD`: denied under default-deny when `bash.exec` actually executed
- `cargo publish --dry-run`: denied
- `c\argo publish --dry-run`: denied
- `tool.search`-only / hallucinated tool-use runs were excluded from pass/fail evidence and tracked separately in #678

Unchecked architecture/docs item:
- `tools/mod.rs` currently exceeds the 15,000-line architecture budget after merging latest `upstream/dev`. This was not introduced by this PR's final cleanup commit and will be handled in a separate follow-up PR.
- Strict docs governance still reports pre-existing baseline docs/release-artifact issues outside the `bash.exec` files changed here. Those issues were not introduced by this PR and will be handled separately rather than silently expanded into this PR.
```

## User-visible / Operator-visible Changes

- `bash.exec` now exists as an experimental discoverable runtime tool when Bash is available.
- `bash.exec` uses AST-governed prefix-rule enforcement rather than executing every command string under the old shell policy model.
- Default Bash rules now live under `~/.loongclaw/rules`, with compatibility inputs still derived from legacy `shell_*` settings.
- Broken governance rules now hide `bash.exec` from discovery instead of advertising a tool that can only fail closed at execution time.

## Failure Recovery

- Fast rollback or disable path:
  - Revert the `bash.exec` governance commits, or make Bash unavailable so `bash.exec` stays hidden.
  - Remove / fix local `~/.loongclaw/rules` files if discovery or execution starts failing closed due to malformed rules.
- Observable failure symptoms reviewers should watch for:
  - `bash.exec` visible when governance rule loading is broken.
  - Commands with escaped static tokens bypassing deny rules.
  - Default rules resolving from the wrong directory instead of `~/.loongclaw/rules`.
  - Models stopping after `tool.search` or hallucinating tool use instead of performing the actual tool call (tracked separately in `#678`).

## Reviewer Focus

- Please focus on:
  - `crates/app/src/tools/bash.rs`
  - `crates/app/src/tools/bash_ast.rs`
  - `crates/app/src/tools/bash_governance.rs`
  - `crates/app/src/tools/bash_rules.rs`
  - `crates/app/src/tools/runtime_config.rs`
  - `crates/app/src/tools/catalog.rs`
  - `crates/app/src/tools/mod.rs`
  - `crates/app/src/config/tools.rs`
  - `README.md`
  - `README.zh-CN.md`
  - `docs/product-specs/tool-surface.md`
- Edge cases worth checking:
  - escaped static tokens such as `c\argo` / `r\m`
  - parse-unreliable or unsupported Bash structures under default-deny
  - home-scoped rules-dir fallback vs explicit `rules_dir`
  - fail-close visibility when governance rules do not load
  - correctness of docs examples versus actual legacy `shell_*` compatibility semantics

## Summary by CodeRabbit

* **New Features**
  * Experimental bash.exec runtime tool: run command/cwd/timeout with conservative AST-backed governance, prefix allow/deny rules, runtime discovery, timeout/output caps, and default non-login-shell; legacy shell_allow/shell_deny are honored for compatibility and rules_dir can be overridden.

* **Bug Fixes**
  * Fixed rules-directory resolution to default to ~/.loongclaw/rules and improved visibility gating so bash.exec is hidden when governance fails.

* **Documentation**
  * Expanded READMEs, examples, plans, and tool-surface docs with concrete rule snippets and guidance.

* **Chores**
  * Updated advisory ignore list for release readiness.

### CODE REVIEW

File: crates/app/src/config/tools.rs [hunk 1]
```diff
@@ -829,6 +842,15 @@ impl ToolConfig {
     }
 }
 
+impl BashToolConfig {
+    pub fn resolved_rules_dir(&self) -> PathBuf {
+        self.rules_dir
+            .as_deref()
+            .map(expand_path)
+            .unwrap_or_else(|| default_loongclaw_home().join("rules"))
+    }
```
[the remark this answers is not in the document]
Comment (ThreeIce) [reply crates/app/src/config/tools.rs:853]: Fixed in 10825f22 and still present in current head. BashToolConfig::resolved_rules_dir now treats blank and whitespace rules_dir as unset before expand_path, then falls back to default_loongclaw_home().join("rules"). Added regression coverage in config::tools::tests::bash_tool_config_treats_blank_rules_dir_override_as_unset.

File: crates/app/src/tools/bash.rs [hunk 2]
```diff
@@ -0,0 +1,357 @@
...
+    if login_shell {
+        vec!["-lc".to_owned(), command.to_owned()]
+    } else {
+        vec!["-c".to_owned(), command.to_owned()]
+    }
+}
+
+pub(super) fn detect_bash_runtime_policy() -> BashExecRuntimePolicy {
+    for candidate in bash_runtime_candidates() {
+        if probe_bash_candidate(&candidate) {
```
[the remark this answers is not in the document]
Comment (ThreeIce) [reply crates/app/src/tools/bash.rs:62]: Fixed in 10825f22. bash runtime probing is now bounded via probe_bash_candidate_with_timeout. Timeout or probe failure returns unavailable and follows fail-closed runtime behavior. Added tools::bash::tests::probe_bash_candidate_with_timeout_returns_false_for_slow_candidate.

File: crates/app/src/tools/bash_rules.rs [hunk 3]
```diff
@@ -0,0 +1,474 @@
...
+    S: AsRef<str>,
+{
+    compile_rules(commands.into_iter().filter_map(|command| {
+        let normalized = command.as_ref().trim().to_ascii_lowercase();
+        if normalized.is_empty() {
+            return None;
+        }
+
+        Some(PrefixRuleSpec {
+            source: format!("{source}:{normalized}"),
```
[the remark this answers is not in the document]
Comment (ThreeIce) [reply crates/app/src/tools/bash_rules.rs]: Fixed in 10825f22 and refined in 41c3beab. Compatibility rules now match using lowered argv tokens only for legacy shell_allow and shell_deny sources, while rule-file rules remain case-sensitive. Added tests compatibility_rules_match_mixed_case_command_tokens and starlark_rules_keep_case_sensitive_matching.

File: crates/app/src/tools/bash_governance.rs [hunk 4]
```diff
@@ -0,0 +1,510 @@
...
+            .map(parse_unreliable_unit_outcome)
+            .collect::<Vec<_>>();
+        return BashGovernanceOutcome {
+            command: command.to_owned(),
+            parse_unreliable: true,
+            unit_outcomes,
+            default_mode,
+            final_decision: final_default_decision(default_mode),
+            final_source: FinalDecisionSource::ParseUnreliableDefault,
+        };
```
[the remark this answers is not in the document]
Comment (ThreeIce) [reply crates/app/src/tools/bash_governance.rs:161]: Fixed in 10825f22. parse_unreliable no longer bypasses evaluate_unit. We now evaluate units first, preserve explicit deny precedence, then apply DefaultParseUnreliable fallback to remaining outcomes. Added tools::bash_governance::tests::parse_unreliable_analysis_preserves_explicit_deny_precedence.

File: crates/app/src/tools/bash_ast.rs [hunk 5]
```diff
@@ -0,0 +1,537 @@
...
+
+    for child in children {
+        let preceding_operator = if let Some(previous_end) = previous_end {
+            let separator = source_text_between(source, previous_end, child.start_byte());
+            if separator_supports_sequential_split(separator) {
+                Some(UnitOperator::Sequential)
+            } else {
+                reliable = false;
+                Some(UnitOperator::Sequential)
+            }
```
[the remark this answers is not in the document]
Comment (ThreeIce) [reply crates/app/src/tools/bash_ast.rs:126]: Fixed in 10825f22. separator_supports_sequential_split now accepts comment spans between sequential units, and collect_program_units plus collect_list_units skip named comment nodes. Added tools::bash_ast::tests::comments_between_sequential_units_do_not_mark_parse_unreliable.

File: crates/app/src/tools/runtime_config.rs [hunk 6]
```diff
@@ -706,6 +846,11 @@ impl ToolRuntimeConfig {
         let config_path = std::env::var("LOONGCLAW_CONFIG_PATH")
             .ok()
             .map(PathBuf::from);
+        let shell_allow: BTreeSet<String> = crate::config::DEFAULT_SHELL_ALLOW
+            .iter()
+            .map(|value| (*value).to_owned())
+            .collect();
+        let shell_deny = BTreeSet::new();
```
[the remark this answers is not in the document]
Comment (ThreeIce) [reply crates/app/src/tools/runtime_config.rs:854]: No code change needed in this PR. Current head already satisfies this contract: from_env keeps the allowlist empty for bash governance defaults, and tools::runtime_config::tests::from_env_defaults_to_empty_allowlist is green on current head.

File: crates/app/src/tools/mod.rs [hunk 7]
```diff
@@ -850,6 +860,7 @@ fn dispatch_tool_request(
             feishu::execute_feishu_tool_with_config(request, config)
         }
         "shell.exec" => shell::execute_shell_tool_with_config(request, config),
+        "bash.exec" => bash::execute_bash_tool_with_config(request, config),
```
[the remark this answers is not in the document]
Comment (ThreeIce) [reply crates/app/src/tools/mod.rs:870]: Fixed in 10825f22. required_capabilities_for_tool_name_and_payload now includes conservative bash.exec classification with InvokeTool, FilesystemRead, FilesystemWrite, and NetworkEgress. Added coverage in required_capabilities_follow_effective_tool_request and tool_search_hides_bash_exec_without_side_effect_capabilities.

File: crates/app/src/tools/mod.rs [hunk 8]
```diff
@@ -63,7 +68,8 @@ pub use catalog::{
     ToolExecutionKind, ToolGovernanceProfile, ToolGovernanceScope, ToolRiskClass,
     ToolSchedulingClass, ToolView, capability_action_class_for_descriptor,
     capability_action_class_for_tool_name, delegate_child_tool_view_for_config,
-    delegate_child_tool_view_for_config_with_delegate, governance_profile_for_descriptor,
+    delegate_child_tool_view_for_config_with_delegate, delegate_child_tool_view_for_runtime_config,
+    delegate_child_tool_view_for_runtime_config_with_delegate, governance_profile_for_descriptor,
```
[the remark this answers is not in the document]
Comment (ThreeIce) [reply crates/app/src/tools/mod.rs:72]: No code change in this PR. Rechecked current head: the runtime-less path that passes None into build_delegate_child_tool_view remains fail-closed for bash.exec because ToolVisibilityGate::BashRuntime requires a runtime-ready policy and None does not satisfy that gate. A full migration to remove runtime-less delegate helpers is reasonable cleanup but out of this PR scope boundary.

File: crates/app/src/tools/catalog.rs [hunk 9]
```diff
@@ -2432,6 +2501,35 @@ fn shell_exec_definition(descriptor: &ToolDescriptor) -> Value {
...
+                        "description": "Bash command to execute."
+                    },
+                    "cwd": {
+                        "type": "string",
+                        "description": "Optional working directory."
+                    },
+                    "timeout_ms": {
+                        "type": "integer",
+                        "description": "Optional command timeout in milliseconds. Defaults to 120000 and is clamped to 1000..=600000."
+                    }
```
[the remark this answers is not in the document]
Comment (ThreeIce) [reply crates/app/src/tools/catalog.rs:2538]: Fixed in 10825f22. The bash.exec catalog schema now sets timeout_ms minimum 1000 and maximum 600000, aligned with runtime clamping in bash.rs. Also extended tools::tests::bash_exec_catalog_exposes_command_cwd_and_timeout_ms.

File: docs/releases/architecture-drift-2026-04.md [hunk 10]
```diff
@@ -23,13 +23,13 @@
...
+| tools_mod | `structural_size` | `crates/app/src/tools/mod.rs` | 15263 | 15000 | -263 | 54 | 70 | 16 | 101.8% | BREACH | 14472 | 5.5% | PASS | 54 |
 | daemon_lib | `structural_size` | `crates/daemon/src/lib.rs` | 6324 | 6500 | 176 | 210 | 210 | 0 | 100.0% | TIGHT | 6324 | 0.0% | PASS | 210 |
 | onboard_cli | `structural_size` | `crates/daemon/src/onboard_cli.rs` | 9519 | 9800 | 281 | 228 | 250 | 22 | 97.1% | TIGHT | 9519 | 0.0% | PASS | 228 |
 
 ## Prioritization Signals
-- BREACH hotspots (>100% of any tracked budget): none
-- TIGHT hotspots (>=95% of any tracked budget): spec_runtime (100.0%), spec_execution (99.4%), acpx_runtime (96.4%), channel_registry (98.9%), channel_config (100.0%), chat_runtime (95.6%), turn_coordinator (96.7%), tools_mod (96.5%), daemon_lib (100.0%), onboard_cli (97.1%)
+- BREACH hotspots (>100% of any tracked budget): tools_mod (101.8%)
+- TIGHT hotspots (>=95% of any tracked budget): spec_runtime (100.0%), spec_execution (99.4%), acpx_runtime (96.4%), channel_registry (98.9%), channel_config (100.0%), chat_runtime (95.6%), turn_coordinator (96.7%), daemon_lib (100.0%), onboard_cli (97.1%)
```
[the remark this answers is not in the document]
Comment (ThreeIce) [reply docs/releases/architecture-drift-2026-04.md]: No code change in this PR by design. The tools_mod BREACH versus release PASS semantics are tracked as a separate follow-up PR per agreed scope boundary. This PR only refreshes the monthly drift artifact.

File: docs/plans/2026-03-29-bash-governance-ast-prefix-rule-implementation-plan.md [hunk 11]
```diff
@@ -0,0 +1,918 @@
...
+## Implementation Notes
+
+- This plan implements [`2026-03-29-bash-governance-ast-prefix-rule-design.md`](2026-03-29-bash-governance-ast-prefix-rule-design.md), not the whole of issue `#637`.
+- `shell.exec` remains untouched in behavior and policy semantics.
+- `approval_required`, `approve_once`, and `approve_always` remain out of scope.
+- Default rules directory behavior:
+  - if `tools.bash.rules_dir` is configured, resolve and use it
+  - otherwise use `.loongclaw/rules` relative to the config file parent when `config_path` is known
+  - otherwise use `.loongclaw/rules` relative to the current working directory
+- Missing rules directory is not an error; broken `.rules` files are an error.
```
[the remark this answers is not in the document]
Comment (ThreeIce) [reply docs/plans/2026-03-29-bash-governance-ast-prefix-rule-implementation-plan.md:48]: Fixed in 10825f22. Updated the historical implementation plan wording to match the shipped default rules directory contract at ~/.loongclaw/rules.

File: docs/plans/2026-04-01-bash-exec-final-cleanup-implementation-plan.md [hunk 12]
```diff
@@ -0,0 +1,421 @@
...
+- `cargo deny check advisories bans sources`
+- `bash scripts/check_architecture_drift_freshness.sh docs/releases/architecture-drift-2026-04.md`
+- `LOONGCLAW_RELEASE_DOCS_STRICT=1 scripts/check-docs.sh`
+- `./scripts/check_dep_graph.sh`
+- `LOONGCLAW_ARCH_STRICT=true ./scripts/check_architecture_boundaries.sh`
+- `git diff --check`
+
+Expected:
+
+- all commands pass on the final tree
```
[the remark this answers is not in the document]
Comment (ThreeIce) [reply docs/plans/2026-04-01-bash-exec-final-cleanup-implementation-plan.md]: Fixed in 10825f22. Updated the final milestone wording so scope-in checks must pass, while the two accepted baseline blockers remain explicitly documented for separate follow-up PRs.

File: docs/plans/2026-03-29-bash-governance-ast-prefix-rule-design.md [hunk 13]
```diff
@@ -0,0 +1,383 @@
...
+    pattern = ["cargo", "test"],
+    decision = "allow",
+    note = "verification command",
+)
+
+prefix_rule(
+    pattern = ["cargo", "publish"],
+    decision = "deny",
+    note = "release path blocked",
+)
```
[the remark this answers is not in the document]
Comment (ThreeIce) [reply docs/plans/2026-03-29-bash-governance-ast-prefix-rule-design.md:102]: Fixed in 10825f22. Updated the historical design plan examples to remove unsupported note fields from prefix_rule samples instead of widening parser scope in this PR.

File: docs/plans/2026-03-27-bash-exec-basic-tool-implementation-plan.md [hunk 14]
```diff
@@ -0,0 +1,841 @@
...
+- `cargo test -p loongclaw-app unavailable_runtime_policy_carries_warning --lib -- --exact --nocapture`
+
+Expected: PASS
+
+- [ ] **Step 6: Commit the runtime-policy slice**
+
+Run:
+
+- `git add crates/app/src/tools/bash.rs crates/app/src/tools/runtime_config.rs`
+- `git commit -m "feat(app): add bash.exec runtime availability policy"`
```
[the remark this answers is not in the document]
Comment (ThreeIce) [reply docs/plans/2026-03-27-bash-exec-basic-tool-implementation-plan.md:243]: Fixed in 10825f22. Updated the historical implementation plan staging list so that runtime-policy slice now includes crates/app/src/config/tools.rs.

File: docs/plans/2026-03-29-bash-governance-ast-prefix-rule-implementation-plan.md [hunk 15]
```diff
@@ -0,0 +1,920 @@
...
+```
+
+- [x] **Step 3: Run the red tests**
+
+Run:
+
+- `cargo test -p loongclaw-app tool_config_parses_bash_rules_dir_override --lib -- --exact --nocapture`
+- `cargo test -p loongclaw-app bash_tool_config_defaults_to_no_explicit_rules_dir --lib -- --exact --nocapture`
+- `cargo test -p loongclaw-app tool_runtime_config_uses_default_workspace_rules_dir_when_unset --lib -- --exact --nocapture`
+- `cargo test -p loongclaw-app tool_runtime_config_projects_bash_rules_dir_override --lib -- --exact --nocapture`
```
[the remark this answers is not in the document]
Comment (ThreeIce) [reply docs/plans/2026-03-29-bash-governance-ast-prefix-rule-implementation-plan.md:191]: Fixed in 41c3beab. Refreshed the historical AST governance plan snippets to current bash_rules API: updated test name, prefix field references, compatibility helper signatures, and merge_rule_sources usage.

File: docs/plans/2026-03-27-bash-exec-basic-tool-implementation-plan.md [hunk 16]
```diff
@@ -0,0 +1,841 @@
...
+
+#[cfg(feature = "tool-shell")]
+#[test]
+fn runtime_tool_view_includes_bash_exec_when_runtime_is_available() {
+    let mut config = runtime_config::ToolRuntimeConfig::default();
+    config.bash_exec = runtime_config::BashExecRuntimePolicy {
+        available: true,
+        command: Some(std::path::PathBuf::from("bash")),
+        warning: None,
+    };
```
[the remark this answers is not in the document]
Comment (ThreeIce) [reply docs/plans/2026-03-27-bash-exec-basic-tool-implementation-plan.md:274]: Fixed in 41c3beab. Updated the historical basic-tool plan snippets so all BashExecRuntimePolicy examples include login_shell.

File: crates/app/src/tools/bash_rules.rs [hunk 17]
```diff
@@ -0,0 +1,528 @@
...
+    if is_legacy_shell_rule(rule) {
+        lowered_command
+    } else {
+        command
+    }
+}
+
+fn is_legacy_shell_rule(rule: &CompiledPrefixRule) -> bool {
+    rule.source.starts_with("shell_allow:") || rule.source.starts_with("shell_deny:")
+}
```
[the remark this answers is not in the document]
Comment (ThreeIce) [reply crates/app/src/tools/bash_rules.rs:205]: Fixed in 41c3beab. Added structured origin metadata on CompiledPrefixRule and switched legacy matching checks to rule.origin instead of source string prefixes. Added tools::bash_rules::tests::rule_file_origin_stays_case_sensitive_even_with_legacy_like_source_prefix.

File: crates/app/src/tools/bash_rules.rs [hunk 18]
```diff
@@ -0,0 +1,528 @@
...
+    S: AsRef<str>,
+{
+    let command: Vec<String> = command
+        .into_iter()
+        .map(|token| token.as_ref().trim().to_owned())
+        .collect();
+    let lowered_command: Vec<String> = command
+        .iter()
+        .map(|token| token.to_ascii_lowercase())
+        .collect();
```
[the remark this answers is not in the document]
Comment (ThreeIce) [reply crates/app/src/tools/bash_rules.rs:167]: Fixed in 41c3beab. Removed token trimming from evaluate_prefix_rules so prefix matching preserves exact decoded argv tokens. Added tools::bash_rules::tests::prefix_rules_preserve_whitespace_in_command_tokens.

File: crates/app/src/tools/mod.rs (hunk 8 above)
[the remark this answers is not in the document]
Comment (ThreeIce) [reply crates/app/src/tools/mod.rs:72]: Verified against current head. This is outside the currently approved convergence slice for this PR. The agreed scope of this follow-up is limited to the unused test-only import fix and the two documentation-alignment fixes captured in docs/plans/2026-04-02-pr677-scope-in-review-convergence-plan.md. Changing the runtime-less delegate-child helpers would expand this PR back into crates/app/src/tools/mod.rs / crates/app/src/tools/catalog.rs behavior work, so this thread is being deferred to a separate follow-up rather than changed in this round.

**Review by chumyin — APPROVED** [review]

**Review by chumyin — APPROVED** [review]

**Comment by chumyin** [comment]
hi @ThreeIce, i tightened the unix bash.exec tests so they no longer depend on the fake runtime path that was tripping ubuntu with `text file busy`, hardened the browser companion request framing and fixture isolation to remove flaky subprocess reads, refreshed the april drift report after rebasing onto latest `dev`, and synced the lockfile so locked ci stays clean on the rebased head. i reran `cargo fmt --all -- --check`, `cargo clippy --workspace --all-targets --all-features -- -D warnings`, and `cargo test --workspace --all-features --locked` locally, and github checks are green now. thanks for sticking with the iteration here, and congrats on getting `bash.exec` landed.
